compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1619,34 +1619,56 @@ pub type PlaceholderConst = Placeholder<BoundVar>;
 /// know their corresponding parameters. We (ab)use this by
 /// calling `type_of(param_did)` for these arguments.
 ///
+/// # Examples
+///
 /// ```rust
 /// #![feature(const_generics)]
 ///
 /// struct A;
 /// impl A {
-///     fn foo<const N: usize>(&self) -> usize { N }
+///     fn foo<const N: usize>(&self) -> [u8; N] { [0; N] }
+///     //           ^ const parameter
 /// }
 /// struct B;
 /// impl B {
-///     fn foo<const N: u8>(&self) -> usize { 42 }
+///     fn foo<const M: u8>(&self) -> usize { 42 }
 /// }
 ///
 /// fn main() {
 ///     let a = A;
-///     a.foo::<7>();
+///     let _b = a.foo::<7>();
+///     //               ^ const argument
 /// }
 /// ```
+///
+/// The important section here is the call `a.foo::<7>()`. We do not know which
+/// `foo` is used until we know the type of `a`.
+///
+/// We only know the type of `a` once we are inside of `typeck(main_def_id)`.
+/// During that we also want to know the type of `_b` which depends on the
+/// the value of the const argument.
+///
+/// To evaluate that const argument we need to know its type, requiring us to
+/// resolve `foo` as it can otherwise be either `usize` of `u8` in this example.
+/// Resolving `foo` once again requires `typeck(main_def_id)` though,
+/// which results in a cycle.
+///
+/// We only care about the const argument once we know which paramter they instantiate.
+/// So instead of using `type_of(const_arg)` we can use `type_of(const_param)` which does
```

Comment:
This is the first time `type_of(const_arg)` comes up, why is it relevant?

Reply:
Oh, it already came up before but not under this name. Please call the same thing by the same name each time it comes up. :)

Reply:
But then... we don't know `type_of(const_param)` either until we have resolved `foo`, so I don't see how this solves anything.

Reply:
```rust
/// not depend on `typeck`. This however requires us to pass the `const_param`
/// to every place which potentionally needs the type of a const argument and gets used
/// during `typeck`.
```
elaborated a bit on that section, I hope that this is now a bit clearer